File: main_project/tripod-new-branch/src/cpp/producermanager/CallFilter.cpp
#include "CallFilter.h"

namespace com {
namespace renren {
namespace tripod {

CountPair::CountPair(int maxCount) : curCount_(0), maxCount_(maxCount) {
}

bool CountPair::inc() {
    IceUtil::Mutex::Lock lock(this);
    return ++curCount_ <= maxCount_;
}

void CountPair::dec() {
    IceUtil::Mutex::Lock lock(this);
    --curCount_;
}

FilterHolder::FilterHolder() {
    //add business config here
}

CallFilter::CallFilter(const FilterHolder& filterHolder, const std::string& businessId) :
                        filterHolder_(filterHolder), businessId_(businessId) {
}

bool CallFilter::filter() {
    std::map<std::string, CountPairPtr>::iterator findIt = filterHolder_.filterMap_.find(businessId_);
    if(findIt == filterHolder_.filterMap_.end()) {
        return false;
    }

    CountPairPtr counter = findIt->second;
    if(counter->inc()) {
        return true;
    }else {
            std::ostringstream os;
            os << "CallFilter::filter too many calls for business " << businessId_ << " max is " << counter->maxCount;
            MCE_WARN(os.str());
            return false;
    }
}

CallFilter::~CallFilter() {
    std::map<std::string, CountPairPtr>::iterator findIt = filterHolder_.filterMap_.find(businessId_);
    if(findIt == filterHolder_.filterMap_.end()) {
        return;
    }

    CountPairPtr counter = findIt->second;

    counter->dec();
}

}
}
}

File: main_project/tripod-new-branch/src/java/TripodProducerManager/src/main/java/com/renren/tripod/CallFilter.java
package com.renren.tripod;

import java.util.HashMap;

import com.xiaonei.xce.log.Tripod;

public class CallFilter {

	private class CountPair {
		private int _curCount = 0;
		private int _maxCount;

		CountPair(int maxCount) {
			_maxCount = maxCount;
		}

		synchronized boolean inc() {
			return ++_curCount >= _maxCount;
		}

		synchronized void dec() {
			--_curCount;
		}
	}

	private class FilterHolder {
		private HashMap<String, CountPair> _filterMap = new HashMap<String, CountPair>();

		public FilterHolder() {
			// add business config here
		}

		public CountPair getCounter(String businessId) {
			return _filterMap.get(businessId);
		}
	}

	FilterHolder _filterHolder = new FilterHolder();

	public boolean filter(String businessId) {
		CountPair cp = _filterHolder.getCounter(businessId);
		if (cp != null) {
			if (cp.inc()) {
				return true;
			} else {
				Tripod.getLogger().error(
						"CallFilter.filter too many call for business "
								+ businessId + " max is " + cp._maxCount);
			}
		}
		return false;
	}

	public void exitFilter(String businessId) {
		CountPair cp = _filterHolder.getCounter(businessId);
		if (cp != null) {
			cp.dec();
		}
	}

}
